### PR TITLE
Fix infinite cors scraper

### PR DESCRIPTION
**Pull request by li-kai**
Uses cheerio and parse5 to scrape cors, without use of regex, the scraper no longer goes into an infinite loop.

### CODE REVIEW

File: tasks/cors.js [hunk 1]
```diff
@@ -7,97 +7,105 @@ module.exports = function (grunt) {
...
-              Department: match[3],
-              CorrectAsAt: /Correct as at ([^<]+)</.exec(data)[1]
+            const $ = cheerio.load(webpage);
+
+            const timestamp = $('h2').text().match(/Correct as at ([^<]+)/).pop();
+
+            // first table consist of details of the module
+            const moduleDetails = $('.tableframe').first().find('tr td:nth-child(2)');
+            const module = {
+              'Type': type,
```
Comment (yangshun) [inline tasks/cors.js]: Nit: Need quotes around the keys?

File: tasks/cors.js [hunk 2]
```diff
@@ -7,97 +7,105 @@ module.exports = function (grunt) {
...
+                return $('td', el).length > 6;
               });
-            });
 
-            callback(null, mod);
+              // get all the relevant information
+              const timetableDetails = rows.map((i, el) => {
+                const row = $('td', el);
+                return {
+                  'ClassNo': row.eq(0).text().trim(),
```
Comment (yangshun) [inline tasks/cors.js]: Same for quotes around keys

File: tasks/helpers.js [hunk 3]
```diff
@@ -46,6 +48,11 @@ exports.requestCached = function (url, options, callback) {
         }
         switch (response.statusCode) {
           case 200:
+            if (!isBinaryFile) {
+              // fix html before saving
+              var doc = parse5.parse(body);
```
Comment (yangshun) [inline tasks/helpers.js]: `const`

File: tasks/helpers.js [hunk 4]
```diff
@@ -36,7 +37,8 @@ exports.requestCached = function (url, options, callback) {
         options.headers = options.headers || {};
         options.headers['if-modified-since'] = (new Date(stats.mtime)).toUTCString();
       }
-      if (isBinaryPath(url)) {
+      var isBinaryFile = isBinaryPath(url);
```
Comment (yangshun) [inline tasks/helpers.js]: `const`

File: tasks/cors.js [hunk 5]
```diff
@@ -7,97 +7,105 @@ module.exports = function (grunt) {
...
-        semester = match[2];
-
-        var urlDeptPattern = /(ModuleD.+)">([^<]+)[\s\S]+?> (.*)<\/div>\s*<\/td>\s*<\/?tr/g;
-        var urlDeptMatches = helpers.matches(urlDeptPattern, data);
-        async.mapSeries(urlDeptMatches, function (match, callback) {
-          helpers.requestCached(options.baseReportUrl + match[1], options, function (err, data) {
+        const $ = cheerio.load(webpage);
+        const listingInfo = $('h2').text().split(':');
+
+        academicYear = listingInfo[1].match(/\d{4}\/\d{4}/).shift();
```
Comment (li-kai) [inline tasks/cors.js]: `academicYear` is used all the way at the bottom for saving to file.

**Pull request merged**